### ISSUE TITLE
Speeding up GUT main scene

### ISSUE THREAD

**Issue by nightblade9** [bitwes/Gut#364]
When I launch my `TestMain.tscn` (which is just an instance of the GUT GUI runner), Godot takes a long time to load - sometimes around ten seconds - presumably while it discovers all the tests. 

In terms of "data shape," I have:
- 40 test classes
- 20 (nested) directories and subdirectories
- Every test uses `preload` to load stuff

Is there anything I can try to speed this up? Maybe switching from `preload` to `load`?

**Comment by bitwes** [comment]
I don't think the directory structure or number of classes is the issue.  I don't have as many directories, but I do have around 45 scripts, some with many inner classes (45 in one of them), with 1100+ tests, and it takes about 2-3 seconds to start (I'm using a 2016 MacBook Pro).

You might be right about using `load` instead of `preload`, unfortunately I think you will have to move all your `load` calls to `before_all`.  `preload` is called when the script itself is loaded (`var Bar = load(res://bar.gd')`).  `load` is called when the script is instantiated (`foo = Bar.new()`).  GUT creates instances of all scripts on startup, so just switching them all to `load` isn't going to save you any time.  If you put them all in `before_all`, then they won't be called until the tests in that script are ran.

Remember, that anything loaded by your test scripts will also trigger `preload` in the loaded scripts.  So if your other game scripts are using `preload` a lot, that could cause some slowdown as well.

Try running GUT with a directory that has a few scripts in it (change config to only use a single directory).  If it is still slow, then there might be something else going on.

Have you tried the new GUT Panel introduced in 7.2.0?  You said you are using a scene to run your tests and I am curious if that is because the panel is missing some functionality.

**Comment by nightblade9** [comment]
- Changing `preload` to `load` had no visible effect on the runtime
- Changing Gut to look at a small subtree instead of the root, made it load very fast
- What GUT Panel did you mean? I don't see any GUT types other than Gut, GdUnitInspector, GdUnitUpdate, and a bunch of descendants of Node (e.g. GdUnitExecutor, GdUnitTestSuite).

**Comment by bitwes** [comment]
I would add more and more scripts to that directory to see if you can find
a threshold where startup takes longer.

GUT panel is the panel added to Godot editor.  Should see "GUT" at the
bottom, next to "Output" and "debug", etc if you are on GUT >= 7.2.0

**Comment by bitwes** [comment]
This is the gut panel.

https://youtu.be/tWNswMIJHKk

**Comment by nightblade9** [comment]
> This is the gut panel.

Nope, didn't try it. 

> Also, did switching make it faster?

I don't know. I think GUT accounts for 80%+ of my tests, and it'll take forever to port them over. I'm moving tests one by one as I touch them (or as I create new tests).

You're probably asking "did drastically reducing the number of GUT tests fix the problem?" and I don't know. Judging by how changing the test root directory to a subdirectory fixed the issue, I would guess that, yes, this will speed up the GUT launch a lot.

**Comment by bitwes** [comment]
Ok.  I'd like to hear some opinions on the GUT Panel.  It makes running subsets of tests through the Godot editor very simple.  I'd also like to know how it compares to running tests using gdUnit3.  There's also some big improvements coming in the next release.  TheDuckCow made a video of him using the first pass of features (currently in master) and I've made more enhancements to it that have not yet been merged:  https://github.com/bitwes/Gut/issues/363#issuecomment-1153279665

**Comment by nightblade9** [comment]
Do you have non-video instructions available somewhere on how to use the panel?

**Comment by bitwes** [comment]
I have not 0ut them together yet, but 8t should be pretty self
explanatory.  All the buttons and settings have mouse over text.

The main thing is that you have to set your test directories in the
settings for the buttons to work (it tells you 8f they aren't set).  Then
you get buttons that will run scripts based on where your cursor is in the
test script.

There are buttons for the full suite, script, inner class, and single test.

And you must have the plugin enabled for the panel to show up.

**Comment by nightblade9** [comment]
I tried this again today. Informally timed (by mouth, not with a device) it seems like launching from the panel, with all tests, is either slightly (20%) slower, or the same as launching `TestMain.tscn`. Same runtime if I specify a single test script file by name under `Script`, and then click `Run Like:`.

Is that what you wanted to know?

**Comment by bitwes** [comment]
You had mentioned that one of the other frameworks made it easier to run
part of the test suite.  I was looking for feedback in that regard.

**Comment by nightblade9** [comment]
> You had mentioned that one of the other frameworks made it easier to run

I haven't tried WAT, albeit I heard good things about it.

I use [gdUnit3](https://github.com/MikeSchulze/gdUnit3/). It supports in my opinion, one key workflow: run a single test, without running any other tests from that class/script. (This is very useful for me in my current project, where I have hard-to-find bugs, and setting breakpoints in common code means they fire too many times during a test suite to be useful.)

The workflow in gdUnit3 is: right-click a test class, pick Debug Tests. That shows a list of tests in Godot, and then you can right-click and run individual tests.

In my opinion, the time GUT takes to start up is "waste" - in gdUnit3, I don't have any time cost to running a single suite or test-case, I just right-click and go. 

Does that make sense?

**Comment by bitwes** [comment]
Can you run GUT from the command line with the `-gh` option and let me know about how long that takes to run?  I've done some basic benchmarking by running that command in a couple different projects of my own.  This will basically just measure the fastest a project can launch and quit.  In my larger projects with a lot of globals and load calls, it takes about 4 seconds.  In a tiny project it takes about 1.

**Comment by nightblade9** [comment]
For the project in question, it takes about ~4 seconds to launch.

**Comment by bitwes** [comment]
Thanks for the info.  I think it should be possible to keep the launch time for a single script to that time if GUT doesn't parse any scripts it doesn't intend to run.

I made a change so that when a script name is specified GUT won't bother loading any script that does not match.  I did not see any performance increase on my end but if the delay is caused by GUT this should help.  Can you test the branch?

https://github.com/bitwes/Gut/tree/i364

**Comment by bitwes** [comment]
I copy/pasted the test suite for GUT 4 times.  Without this change it took 4.877s to run a quick test (GUT reported a 0.5s runtime).  With the change it took 1.323s.  For reference it takes me 1.075s to run `-gh` in the GUT repo.

All times generated using `zsh` and `time` on 2016 mbp.
```
time (godot -s addons/gut/gut_cmdln.gd -gh)
time (godot -s addons/gut/gut_cmdln.gd -gselect=test_spy.gd)
```

**Comment by nightblade9** [comment]
It takes about a second to launch. Interestingly, I launched it thrice; the first launch took the longest (roughly 2s) and the other two launched very quickly.

**Comment by bitwes** [comment]
That's great news.  I'll put this change in the next release.  Thanks a lot for helping work through the issue.

**Comment by bitwes** [comment]
Released in 7.4.0
